lib/Mojolicious/Guides/Testing.pod

```diff
@@ -0,0 +1,802 @@
+    ->json_is('/message' => 'this is five');
+
+When we set C<max_redirects>, it stays set for the life of the test object
+until we change it.
+
+L<Test::Mojo>'s handling of HTTP redirects eliminates the need for making
+many, sometimes an unknown number, of redirections to keep testing precise and
+easy to follow (ahem).
+
+=head2 Cookies and session management
+
+We can use L<Test::Mojo> to test applications that keep session state in
+cookies. By default, the L<Mojo::UserAgent> object in L<Test::Mojo> will
+manage session for us by saving and sending cookies automatically, just like
+common web browsers:
+
+  use Mojo::Base -strict;
+
+  use Test::Mojo;
+  use Test::More;
+
+  my $t = Test::Mojo->new('MyApp');
+
+  # No authorization cookie
+  $t->get_ok('/')
+    ->status_is(401)
+    ->content_is('Please log in');
+
+  # Application sets an authorization cookie
+  $t->post_ok('/login', { password => 'let me in' })
+    ->status_is(200)
+    ->content_is('You are logged in');
+
+  # Sends the cookie from the previous transaction
+  $t->get_ok('/')
+    ->status_is(200)
+    ->content_like(qr/You logged in at \d+/);
+
+  # Clear the cookies
+  $t->reset_session;
+
+  # No authorization cookie again
+  $t->get_ok('/')
+    ->status_is(401)
+    ->content_is('Please log in');
+
+We can also inspect cookies in responses for special values through the
+transaction's response (L<Mojo::Message::Response>) object:
+
+  $t->get_ok('/');
+
+  my ($cookie) = grep { $_->name eq 'smarty' } @{$t->tx->res->cookies};
+  like($cookie && $cookie->to_string, qr(smarty=pants), 'cookie found');
+
+=head2 Custom transactions
+
+Let's say we have an application that responds to a new HTTP verb C<RING> and
+to use it we must also pass in a secret cookie value. This is not a
+problem. We can test the application by creating a L<Mojo::Transaction>
+object, setting the cookie, then passing the transaction object to
+C<request_ok>:
+
+  # Use custom "RING" verb
+  my $tx = $t->ua->build_tx(RING => '/doorbell');
+
+  # Set a special cookie
+  $tx->req->cookies({Secret => "don't tell anybody"});
+
+  # Make the request
+  $t->request_ok($tx)
+    ->status_is(200)
+    ->json_is('/status' => 'ding dong');
+
+=head2 Testing WebSocket web services
+
+While the message flow on WebSocket connections can be rather dynamic, it more
+often than not is quite predictable, which allows this rather pleasant
+L<Test::Mojo> API to be used.
+
+  use Mojo::Base -strict;
+
+  use Test::More;
+  use Test::Mojo;
+
+  # Include application
+  use FindBin;
+  require "$FindBin::Bin/../echo.pl";
+
+  # Test echo web service
+  my $t = Test::Mojo->new;
+  $t->websocket_ok('/echo')
+    ->send_ok('Hello Mojo!')
+    ->message_ok
+    ->message_is('echo: Hello Mojo!')
+    ->finish_ok;
+
+  # Test JSON web service
+  $t->websocket_ok('/echo.json')
+    ->send_ok({json => {test => [1, 2, 3]}})
+    ->message_ok
+    ->json_message_is('/test', [1, 2, 3])
+    ->finish_ok;
+
+  done_testing();
```

Comment:
Might be worth elaborating a little on what stages the WebSocket connection goes through in this section. It feels a little rushed when compared to the rest of the guide.